Add roles to Person WCIF and set up /wcif/persons endpoint

jonatanklosko opened the pull request:
Implements another part of WCIF spec.

### Roles

I added `roles` field to `registrations` table. The user of the API can store any roles here (e.g. *scrambler*, *judge*, *dataentry*, *runner*). Two special roles (implicitly specified in the WCIF spec) being *delegate* and *organizer* are not stored in the `roles` column, they are computed on the fly instead (as they shouldn't be changed using this API).
*Drawback: there is no way to save other roles for registrationless person at the moment (e.g. a delegate that doesn't compete). Still, I think registrations is the right place and we will likely handle those cases in the future (e.g. we talked about registrations without events for such people).*

### /wcif/persons API

As discussed over Slack, the idea for this endpoint is to update people information (`roles` and `assignments` as per the current WCIF spec). The request may contain full people WCIF or partial ones (just `wcaUserId` and the updated field). There should be no way to update immutable data (like name), just the fields specified above.

jonatanklosko commented on WcaOnRails/app/models/user.rb:
```diff
@@ -727,24 +727,26 @@ def serializable_hash(options = nil)
       "registrantId" => registrant_id,
       "countryIso2" => country_iso2,
-      "delegatesCompetition" => competition.delegates.include?(self),
-      "organizesCompetition" => competition.organizers.include?(self),
```
Do we know of any software using this already?

timreyn replied:
CubingUSA does: https://github.com/cubingusa/org/blob/3441353f06f2a1eb85faaa80fefd92acc3d43b64/src/scheduling/wcif/person.py#L53

But it would be easy enough to switch to using the new form.

timreyn replied:
Nah, feel free to just delete it, just please let me know when you push to prod and I'll update.

I'm not worried about a short-term breakage, since it's pretty much just us nats that uses it now.

jonatanklosko replied:
@timreyn I've just recalled that I should've let you know once this gets merged, so doing it now =)

viroulep commented on WcaOnRails/app/controllers/api/v0/competitions_controller.rb:
```diff
@@ -56,6 +56,21 @@ def update_events_from_wcif
     }
   end
 
+  def update_persons_from_wcif
```
I had to implement such an endpoint for the schedule too, and ended up DRYing up the code for the exception handling, see [here](https://github.com/viroulep/worldcubeassociation.org/commit/8aa06b15d9b58a838412a40ae66642da2add9595).

I'm still not done with the test unfortunately, so your PR is likely to be ready faster than mine...
If you think it's a good idea maybe you could take this change right away?

Isn't there a JSON schema validation for the persons WCIF?

viroulep replied:
I'm totally with fixing the conflict myself so if you can just include the changes made to the API competitions controller (and use it of course :p) that's awesome (L39-61 [here](https://github.com/viroulep/worldcubeassociation.org/blob/8aa06b15d9b58a838412a40ae66642da2add9595/WcaOnRails/app/controllers/api/v0/competitions_controller.rb#L39-L61))!
This way I'll just drop the changes to the file and add the `update_schedule_from_wcif` endpoint on top of it.

jonatanklosko replied:
Sounds good, thanks!

jonatanklosko commented:
I added the JSON Schema definition for `persons` and applied the refactoring of controller actions from @viroulep (I also moved the data extraction from `params["_json"]` to the generic method) =)

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -920,6 +920,25 @@ def set_wcif_events!(wcif_events, current_user)
+
+    ActiveRecord::Base.transaction do
+      wcif_persons.each do |wcif_person|
+        # registration = registrations.joins(:user).where("users.id = ?", wcif_person["wcaUserId"]).first
```
Vestigial?

jonatanklosko replied:
Yeah, I did even figure out a better query than both this and the one below.

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -920,6 +920,25 @@ def set_wcif_events!(wcif_events, current_user)
+        # registration = registrations.joins(:user).where("users.id = ?", wcif_person["wcaUserId"]).first
+        user = User.includes(:registrations).find(wcif_person["wcaUserId"])
+        registration = user.registrations.find_by_competition_id(self.id)
+        if registration && wcif_person["roles"]
```
Could you explain the reasoning behind this? It seems to me that we might want to error out if a registration that the user is trying to update doesn't exist, rather than throwing an error.

If this is the right decision, could you add a comment why this is the right thing to do?

jfly replied:
Ahh, I think this is definitely worth a comment, then. If we have a registration-less organizer or delegate, we won't be able to give them new roles, which might be desirable.

I think this is yet another reason why we should support registrations for 0 events.

jonatanklosko replied:
Yeah, that's what I stated in PR description =)

jonatanklosko commented:
Updated!

jonatanklosko commented:
I plan to merge it on Tuesday unless there are any objections. @jfly has already approved it, but I applied some changes as a response to his comments (straightforward stuff).
